File: application/src/components/Forms/src/com/eas/client/forms/api/VerticalPosition.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.eas.client.forms.api;

import com.eas.script.AlreadyPublishedException;
import com.eas.script.HasPublished;
import com.eas.script.NoPublisherException;
import com.eas.script.ScriptFunction;
import com.eas.script.ScriptObj;
import jdk.nashorn.api.scripting.JSObject;

/**
 *
 * @author mg
 */
@ScriptObj(jsDoc = "/**\n"
        + "* Vertical position constants.\n"
        + "*/")
public class VerticalPosition implements HasPublished {

    public static final int TOP = 1;
    public static final int CENTER = 0;
    public static final int BOTTOM = 3;
    
    private static JSObject publisher;
    protected Object published;
    
    @ScriptFunction()
    public static int getTOP() {
        return TOP;
    }
    
    @ScriptFunction
    public static int getCENTER() {
        return CENTER;
    }
    
    @ScriptFunction
    public static int getBOTTOM() {
        return BOTTOM;
    }
    
     @Override
    public Object getPublished() {
        if (published == null) {
            if (publisher == null || !publisher.isFunction()) {
                throw new NoPublisherException();
            }
            published = publisher.call(null, new Object[]{});
        }
        return published;
    }

    @Override
    public void setPublished(Object aValue) {
        if (published != null) {
            throw new AlreadyPublishedException();
        }
        published = aValue;
    }

    public static void setPublisher(JSObject aPublisher) {
        publisher = aPublisher;
    }
}

File: application/api/forms/vertical-position.js
(function() {
    var javaClass = Java.type("com.eas.client.forms.api.VerticalPosition");
    javaClass.setPublisher(function(aDelegate) {
        return new P.VerticalPosition(aDelegate);
    });
    
    /**
     * Generated constructor.
     * @namespace VerticalPosition
     */
    P.VerticalPosition = function () {

        var maxArgs = 0;
        var delegate = arguments.length > maxArgs ?
            arguments[maxArgs] : new javaClass();

        Object.defineProperty(this, "unwrap", {
            get: function() {
                return function() {
                    return delegate;
                };
            }
        });
        /**
         * Generated property jsDoc.
         * @property CENTER
         * @memberOf VerticalPosition
         */
        Object.defineProperty(this, "CENTER", {
            get: function() {
                var value = delegate.CENTER;
                return P.boxAsJs(value);
            }
        });

        /**
         * Generated property jsDoc.
         * @property TOP
         * @memberOf VerticalPosition
         */
        Object.defineProperty(this, "TOP", {
            get: function() {
                var value = delegate.TOP;
                return P.boxAsJs(value);
            }
        });

        /**
         * Generated property jsDoc.
         * @property BOTTOM
         * @memberOf VerticalPosition
         */
        Object.defineProperty(this, "BOTTOM", {
            get: function() {
                var value = delegate.BOTTOM;
                return P.boxAsJs(value);
            }
        });


        delegate.setPublished(this);
    };
})();